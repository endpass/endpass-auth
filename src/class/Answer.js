export default class Answer {
  static createOk(payload) {
    return {
      status: true,
      payload,
    };
  }

  // TODO: `error` field support < 0.24 version
  static createFail(code, error) {
    return {
      status: false,
      code,
      error,
    };
  }
}
