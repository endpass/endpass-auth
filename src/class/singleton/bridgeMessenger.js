import CrossWindowMessenger from '@endpass/class/CrossWindowMessenger';
import { DIRECTION } from '@/constants';

const bridgeMessenger = new CrossWindowMessenger({
  // showLogs: !ENV.VUE_APP_IS_PRODUCTION,
  name: 'auth-bridge',
  target: window.parent,
  from: DIRECTION.AUTH,
  to: DIRECTION.CONNECT,
});

export default bridgeMessenger;
