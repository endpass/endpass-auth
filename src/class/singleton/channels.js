import Channel from '@/class/Channel';

export const authChannel = new Channel('auth');
export const signChannel = new Channel('sign');
export const accountChannel = new Channel('account');
export const permissionChannel = new Channel('permission');

export default {
  authChannel,
  signChannel,
  accountChannel,
  permissionChannel,
};
