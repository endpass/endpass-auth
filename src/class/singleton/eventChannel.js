import EventChannel from '@/class/EventChannel';

export default new EventChannel();
