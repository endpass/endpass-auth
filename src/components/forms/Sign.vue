<template>
  <form
    data-test="sign-form"
    @submit.prevent="emitSubmit"
  >
    <form-field v-if="requesterUrl">
      <a
        :href="requesterUrl"
        data-test="requester-url"
      >{{ requesterUrl }}</a>
      {{ $t('components.sign.requestSign') }}
    </form-field>
    <form-field :label="$t('components.sign.requiresSignBy')">
      <message
        :ellipsis="true"
        data-test="account-address"
      >
        {{ account }}
      </message>
    </form-field>
    <!-- <form-field v-if="error">
      <message
        :error="true"
        data-test="error-message"
      >
        {{ error }}
      </message>
    </form-field> -->
    <form-field :label="$t('components.sign.yourPass')">
      <v-input
        v-model="password"
        :placeholder="$t('components.sign.enterPass')"
        :error="error"
        type="password"
      />
    </form-field>
    <template v-if="isTransaction">
      <form-field label="Value">
        <v-input
          v-model="value"
          :disabled="true"
          type="number"
        />
      </form-field>
      <form-field
        v-if="data"
        :label="$t('components.sign.transactionData')"
      >
        <v-input
          v-model="data"
          :disabled="true"
        />
      </form-field>
      <form-field :label="$t('components.sign.transactionGasPrice')">
        <v-input
          v-model="gasPrice"
          type="number"
        />
      </form-field>
      <form-field v-if="labeledGasPricesList">
        <v-content-switcher
          v-model="gasPrice"
          :items="labeledGasPricesList"
          step="0.0000000001"
          min="0"
        />
      </form-field>
      <form-field :label="$t('components.sign.transactionGasLimit')">
        <v-input
          v-model="gasLimit"
          type="number"
          min="0"
        />
      </form-field>
    </template>
    <form-field
      v-else-if="message"
      :label="$t('components.sign.message')"
    >
      {{ message }}
    </form-field>
    <form-controls>
      <v-button
        :disabled="loading || !password"
        :submit="true"
        type="primary"
        data-test="submit-button"
      >
        {{ primaryButtonLabel }}
      </v-button>
      <v-button
        :disabled="!closable || loading"
        skin="quaternary"
        data-test="cancel-button"
        @click="emitCancel"
      >
        {{ $t('global.close') }}
      </v-button>
    </form-controls>
  </form>
</template>

<script>
import Web3 from 'web3';
import get from 'lodash/get';
import { mapGetters } from 'vuex';
import VInput from '@endpass/ui/kit/VInput';
import VButton from '@endpass/ui/kit/VButton';
import VContentSwitcher from '@endpass/ui/kit/VContentSwitcher';
import Message from '@/components/common/Message.vue';
import FormField from '@/components/common/FormField.vue';
import FormControls from '@/components/common/FormControls.vue';

const { hexToUtf8, fromWei } = Web3.utils;

export default {
  name: 'SignForm',

  props: {
    loading: {
      type: Boolean,
      default: false,
    },

    error: {
      type: String,
      default: null,
    },

    request: {
      type: Object,
      default: null,
    },

    closable: {
      type: Boolean,
      default: true,
    },

    isTransaction: {
      type: Boolean,
      default: false,
    },
  },

  data: () => ({
    password: '',
    value: 0,
    data: '',
    gasPrice: 0,
    gasLimit: 21000,
  }),

  computed: {
    ...mapGetters(['labeledGasPricesList']),

    account() {
      return get(this.request, 'address');
    },

    requesterUrl() {
      return get(this.request, 'url');
    },

    message() {
      const hexMessage = get(this.request, 'request.params[1]');

      if (!hexMessage) return null;

      return hexToUtf8(hexMessage);
    },

    transaction() {
      return get(this.request, 'request.params[0]');
    },

    primaryButtonLabel() {
      return !this.loading
        ? this.$i18n.t('global.sign')
        : this.$i18n.t('global.loading');
    },
  },

  methods: {
    emitSubmit() {
      if (!this.password) return;

      if (!this.isTransaction) {
        this.$emit('submit', {
          account: this.account,
          password: this.password,
        });
        return;
      }

      const updatedTransaction = {
        ...this.transaction,
        value: this.value,
        gasPrice: this.gasPrice,
        gasLimit: this.gasLimit,
      };

      if (this.data) {
        Object.assign(updatedTransaction, {
          data: this.data,
        });
      }

      this.$emit('submit', {
        account: this.account,
        password: this.password,
        transaction: updatedTransaction,
      });
    },

    emitCancel() {
      this.$emit('cancel');
    },
  },

  mounted() {
    if (!this.isTransaction) return;

    this.value = fromWei(this.transaction.value);
    this.gasPrice = fromWei(this.transaction.gasPrice);
    this.data = this.transaction.data;
  },

  components: {
    VButton,
    VInput,
    VContentSwitcher,
    Message,
    FormField,
    FormControls,
  },
};
</script>
