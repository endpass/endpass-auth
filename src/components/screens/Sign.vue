<template>
  <screen @close="handleWindowClose">
    <v-frame
      :loading="!request"
      :closable="isDialog"
      @close="handleSignCancel"
    >
      <sign-form
        :loading="loading"
        :request="request"
        :error="error"
        :closable="isDialog"
        @cancel="handleSignCancel"
        @submit="handleSignSubmit"
      />
    </v-frame>
  </screen>
</template>

<script>
import { mapActions, mapGetters, mapState } from 'vuex';
import Screen from '@/components/common/Screen';
import VFrame from '@/components/common/VFrame';
import SignForm from '@/components/forms/Sign';

export default {
  name: 'Sign',

  data: () => ({
    error: null,
  }),

  computed: {
    ...mapState({
      inited: state => state.core.inited,
      loading: state => state.core.loading,
      request: state => state.requests.request,
    }),
    ...mapGetters(['isDialog']),
  },

  methods: {
    ...mapActions(['processRequest', 'cancelRequest', 'dialogClose']),

    async handleSignSubmit(res) {
      try {
        await this.processRequest(res.password);
        this.error = null;
      } catch (err) {
        console.log(err);
        this.error = err.message;
      }
    },

    handleSignCancel() {
      this.cancelRequest();
      this.dialogClose();
    },

    handleWindowClose() {
      this.cancelRequest();
      this.dialogClose();
    },
  },

  components: {
    Screen,
    SignForm,
    VFrame,
  },
};
</script>
