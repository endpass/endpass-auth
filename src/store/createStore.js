// @ts-check
import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

/**
 * @param {object?} [options] passed options for store
 * @return {import('vuex').Store<{}>} store Vuex store instance
 */
export default options => {
  return new Vuex.Store({
    strict: !ENV.VUE_APP_IS_PRODUCTION,
    ...options,
  });
}
