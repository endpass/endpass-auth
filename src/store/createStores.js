import GasPriceModule from '@/store/modules/GasPriceModule';
import SharedModule from '@/store/modules/SharedModule';
import AccountsModule from '@/store/modules/AccountsModule';
import CoreModule from '@/store/modules/CoreModule';
import RequestsModule from '@/store/modules/RequestsModule';
import createStoreModule from '@/store/createStoreModule';

export default store => {
  const createModule = createStoreModule(store);

  const gasPriceStore = createModule(GasPriceModule, 'gasPrice');

  const sharedStore = createModule(SharedModule, 'shared');

  const accountsStore = createModule(AccountsModule, 'accounts', {
    sharedStore,
  });

  const coreStore = createModule(CoreModule, 'core', {
    accountsStore,
    sharedStore,
  });

  const requestStore = createModule(RequestsModule, 'requests', {
    accountsStore,
    sharedStore,
  });

  return {
    gasPriceStore,
    sharedStore,
    accountsStore,
    coreStore,
    requestStore,
  }
}
