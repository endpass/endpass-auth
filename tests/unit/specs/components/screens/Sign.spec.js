import Vuex from 'vuex';
import { shallowMount, createLocalVue } from '@vue/test-utils';
import Sign from '@/components/screens/Sign';
import setupI18n from '@/locales/i18nSetup';
import { signChannel } from '@/class/singleton/channels';
import Answer from '@/class/Answer';
import createStore from '@/store/createStore';
import createStores from '@/store/createStores';

const localVue = createLocalVue();

localVue.use(Vuex);
const i18n = setupI18n(localVue);

describe('Sign', () => {

  const createWrapper = options => {
    const store = createStore();
    const { accountsStore, requestStore, coreStore } = createStores(store);

    return shallowMount(Sign, {
      accountsStore,
      requestStore,
      coreStore,
      localVue,
      i18n,
      ...options,
    });
  };

  describe('render', () => {
    let wrapper;
    const request = {
      address: '0x0',
      request: {
        id: 1,
        params: ['foo', 'bar'],
        jsonrpc: '2.0',
      },
    };

    beforeEach(() => {
      wrapper = createWrapper();
    });

    describe('render', () => {
      it('should correctly render Sign component', () => {
        requestStore.setRequest(request);
        expect(wrapper.name()).toBe('Sign');
        expect(wrapper.html()).toMatchSnapshot();
      });
    });

    it('should render sign message form if request does not contains transaction', () => {
      requestStore.setRequest({
        request: {
          method: 'eth_sign',
        },
      });
      wrapper = createWrapper();

      expect(wrapper.find('sign-message-form-stub').exists()).toBe(true);
      expect(wrapper.find('sign-transaction-form-stub').exists()).toBe(false);
    });

    it('should render sign transaction form if request contains transaction', () => {
      requestStore.setRequest({
        request: {
          method: 'eth_sendTransaction',
        },
      });
      wrapper = createWrapper();

      expect(wrapper.find('sign-transaction-form-stub').exists()).toBe(true);
      expect(wrapper.find('sign-message-form-stub').exists()).toBe(false);
    });

    describe('behavior', () => {
      it('should confirm current request with password', async () => {
        expect.assertions(1);

        requestStore.setRequest(request);
        const dataPromise = signChannel.take();
        wrapper.find('sign-message-form-stub').vm.$emit('submit', {
          password: 'foo',
        });

        await global.flushPromises();
        const res = await dataPromise;

        expect(res).toEqual(
          Answer.createOk({
            id: 1,
            jsonrpc: '2.0',
            result: 'signature',
          }),
        );
      });

      it('should cancel current request', async () => {
        expect.assertions(1);

        requestStore.setRequest(request);
        const dataPromise = signChannel.take();

        wrapper.find('sign-message-form-stub').vm.$emit('cancel', {
          password: 'foo',
        });

        await global.flushPromises();
        const res = await dataPromise;

        expect(res).toEqual(
          Answer.createOk({
            id: 1,
            error: 'canceled',
            result: [],
          }),
        );
      });
    });
  });
});
