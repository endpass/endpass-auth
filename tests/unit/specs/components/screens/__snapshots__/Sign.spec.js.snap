// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign render render should correctly render Sign component 1`] = `
<screen-stub>
  <v-modal-card-stub isclosable="true">
    <sign-message-form-stub request="[object Object]" closable="true"></sign-message-form-stub>
  </v-modal-card-stub>
</screen-stub>
`;
